File: backend/src/models/PledgeAllocation.js
'use strict';

const { Model, DataTypes } = require('sequelize');

const SOURCES = ['stripe_auto', 'treasurer_manual', 'migration', 'stripe_refund'];

module.exports = (sequelize) => {
  // APPEND-ONLY. Rows are never updated or deleted — corrections are reversing
  // rows (negative amount + reverses_allocation_id + reason). That is what makes
  // the audit trail (who/when/previous/new/why) fall out of the table itself.
  // A Postgres trigger enforces this in production; the service layer must never
  // issue UPDATE or DELETE against this model.
  class PledgeAllocation extends Model {
    static associate(models) {
      PledgeAllocation.belongsTo(models.Pledge, { foreignKey: 'pledge_id', as: 'pledge' });
      PledgeAllocation.belongsTo(models.Transaction, { foreignKey: 'transaction_id', as: 'transaction' });
      PledgeAllocation.belongsTo(models.Member, { foreignKey: 'allocated_by', as: 'allocator' });
      PledgeAllocation.belongsTo(models.PledgeAllocation, {
        foreignKey: 'reverses_allocation_id', as: 'reverses'
      });
    }
  }

  PledgeAllocation.init({
    id: { type: DataTypes.BIGINT, primaryKey: true, autoIncrement: true },
    pledge_id: {
      type: DataTypes.BIGINT, allowNull: false,
      references: { model: 'pledges', key: 'id' }
    },
    transaction_id: {
      type: DataTypes.BIGINT, allowNull: false,
      references: { model: 'transactions', key: 'id' }
    },
    amount: {
      type: DataTypes.DECIMAL(10, 2),
      allowNull: false,
      validate: {
        notZero(value) {
          if (parseFloat(value) === 0) throw new Error('Allocation amount cannot be zero');
        }
      }
    },
    source: {
      type: DataTypes.STRING(24), allowNull: false,
      validate: { isIn: [SOURCES] }
    },
    allocated_by: {
      type: DataTypes.BIGINT, allowNull: true,
      references: { model: 'members', key: 'id' },
      comment: 'NULL only for automated sources (stripe_auto, stripe_refund)'
    },
    reason: { type: DataTypes.TEXT, allowNull: true },
    reverses_allocation_id: {
      type: DataTypes.BIGINT, allowNull: true,
      references: { model: 'pledge_allocations', key: 'id' }
    },
    idempotency_key: {
      type: DataTypes.STRING(191), allowNull: true, unique: true,
      comment: 'auto:<external_id>:<pledge_id> for automated allocations; NULL for manual'
    }
  }, {
    sequelize,
    modelName: 'PledgeAllocation',
    tableName: 'pledge_allocations',
    underscored: true,
    timestamps: true,
    createdAt: 'created_at',
    updatedAt: false,  // rows never change
    indexes: [
      { fields: ['pledge_id'] },
      { fields: ['transaction_id'] },
      { unique: true, fields: ['idempotency_key'] }
    ],
    validate: {
      reversalsAreNegativeAndExplained() {
        if (this.reverses_allocation_id == null) return;
        if (parseFloat(this.amount) >= 0) {
          throw new Error('A reversing allocation must have a negative amount');
        }
        if (!this.reason || !String(this.reason).trim()) {
          throw new Error('A reversing allocation must have a reason');
        }
      }
    }
  });

  PledgeAllocation.SOURCES = SOURCES;

  return PledgeAllocation;
};

File: backendJava/src/main/java/church/abunearegawi/backend/model/PledgeAllocation.java
package church.abunearegawi.backend.model;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;

import java.math.BigDecimal;
import java.time.OffsetDateTime;

/**
 * A payment, or part of one, credited against a pledge.
 *
 * <p><strong>Append-only.</strong> A Postgres trigger blocks UPDATE and DELETE
 * on this table. A mistake is corrected by inserting a negative row that
 * reverses the original and carries a mandatory reason, so the record of what
 * was believed at the time survives the correction. Any Java code that tries to
 * mutate or remove a row will fail at the database, by design — do not add a
 * setter-driven update path here.
 *
 * <p>Fulfilment is derived by summing these, never stored on the pledge. That
 * is why {@code pledges.legacy_status} is frozen: a stored status can disagree
 * with the money, a sum cannot.
 */
@Entity
@Table(name = "pledge_allocations")
@Getter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class PledgeAllocation {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "pledge_id", nullable = false)
    private Long pledgeId;

    @Column(name = "transaction_id", nullable = false)
    private Long transactionId;

    /**
     * Negative on a reversing row. Not constrained to be positive here on
     * purpose — that is the mechanism corrections rely on.
     */
    @Column(nullable = false, precision = 12, scale = 2)
    private BigDecimal amount;

    /**
     * Where the credit came from: the automatic path after an online payment,
     * or a treasurer recording one. Widening this is a schema change — the
     * column carries a CHECK constraint the Node migrations own, which is what
     * Zelle, bank and Square allocation would each need extending.
     */
    @Column(nullable = false, length = 32)
    private String source;

    /** The only values the source CHECK constraint permits. */
    public static final String SOURCE_STRIPE_AUTO = "stripe_auto";
    public static final String SOURCE_TREASURER_MANUAL = "treasurer_manual";
    public static final String SOURCE_MIGRATION = "migration";
    public static final String SOURCE_STRIPE_REFUND = "stripe_refund";

    /** The member who recorded it; null for allocations the system made. */
    @Column(name = "allocated_by")
    private Long allocatedBy;

    /** Mandatory in practice for a reversal: why the correction was made. */
    @Column(columnDefinition = "TEXT")
    private String reason;

    /** Set on a reversing row, pointing at the row it cancels. */
    @Column(name = "reverses_allocation_id")
    private Long reversesAllocationId;

    /**
     * Guards against a webhook delivered twice creating the credit twice.
     * Stripe redelivers on its own schedule, so this is load-bearing rather
     * than defensive.
     */
    @Column(name = "idempotency_key")
    private String idempotencyKey;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private OffsetDateTime createdAt;

    /** A reversing row rather than an original credit. */
    public boolean isReversal() {
        return reversesAllocationId != null;
    }
}
